test/functional/p2p_ping.py

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test ping message
+"""
+
+import time
+
+from test_framework.messages import (
+    msg_pong,
+)
+from test_framework.mininode import (
+    P2PInterface,
+    wait_until,
+)
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal
+
+PING_INTERVAL = 2 * 60
+
+
+class msg_pong_corrupt(msg_pong):
+    def serialize(self):
+        return b""
+
+
+class NodePongAdd1(P2PInterface):
+    def on_ping(self, message):
+        self.send_message(msg_pong(message.nonce + 1))
+
+
+class NodeNoPong(P2PInterface):
+    def on_ping(self, message):
+        pass
+
+
+class PingPongTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
```

Comment:
Is a clean chain needed for p2p testing here? The tests seem to be fine with this set to the default of false.

Reply:
Sending a ping/pong shouldn't require any blocks, so I set this to false for now. This can be removed when the test is extended and needs blocks.

Reply:
Did you still want to set this to false? :angel:

Reply:
Seems overly expensive to copy the blockchain when it is not needed, no?

Reply:
This might be a wrong conception I've had for a long time. I thought it was cheaper to use the cached data directories rather than set up a clean chain, and that the test framework defaults to `self.setup_clean_chain = False` for this reason. I reckoned the default was the cheaper option and the more expensive one was opt-in.

Reply:
It might be a good idea to add a sentence on performance in the "General test-writing advice" section in `functional/README.md` where it states "Set the `self.setup_clean_chain` variable in `set_test_params()` to control whether or not to use the cached data directories."

Reply:
The reason this default to a chain is that most tests require a chain, so so the least verbose option is the default.

Also, defaulting to a chain prevents test writers from accidentally mining a chain when it is not needed. Mining is orders of magnitude slower than copying.